File: packages/js/src/core/codec/nrziCodec.ts
export function nrziEncode(bits: number[]): number[] {
  const out: number[] = [];
  let level = 1;
  for (const bit of bits) {
    if (bit === 0) {
      level ^= 1;
    }
    out.push(level);
  }
  return out;
}

export function nrziDecode(tones: number[]): number[] {
  if (tones.length === 0) {
    return [];
  }
  const out: number[] = [];
  let prev = tones[0];
  for (const tone of tones) {
    out.push(tone === prev ? 1 : 0);
    prev = tone;
  }
  return out;
}

File: packages/python/qraudio/codec/nrziCodec.py
from __future__ import annotations


def nrziEncode(bits: list[int]) -> list[int]:
    out: list[int] = []
    level = 1
    for bit in bits:
        if bit == 0:
            level ^= 1
        out.append(level)
    return out


def nrziDecode(tones: list[int]) -> list[int]:
    if not tones:
        return []
    out: list[int] = []
    prev = tones[0]
    for tone in tones:
        out.append(1 if tone == prev else 0)
        prev = tone
    return out
